File: 2021/day-16/src/main.rs
use day_16::*;

// use bit_field::{BitField};
// use bitvec::prelude::*;



fn main() {
    let input = include_str!("input");

    dbg!(assert_eq!(part_one(input), 0));
    // dbg!(assert_eq!(part_two(input), 2995));
}

fn part_one(input: &str) -> usize {
    println!("Input: {}", input);

    let mut versions: Vec<usize> = vec![];
    let input = BinaryString::from_hex(input);

    // TODO: Use extract_packets

    versions.iter().sum()
}

// fn parse_header(input: &str) -> (usize, usize) {
//     let header = parse_hex(&input[0..2]);

//     (header.get_bits(5..8), header.get_bits(2..5))
// }

// fn parse_literal(input: &str) -> usize {
//     let literal_start = input.len() * 4 - 6;
    
//     let parsed = parse_hex(input);
//     let literal_bits = BitSlice::<Msb0, _>::from_element(&parsed);
    
//     let mut result = BitVec::new();

//     for group in literal_bits[literal_bits.len()-literal_start..].chunks(5) {
//         let literal = &group[1..];

//         result.extend(literal);

//         if !group[0] {
//             break
//         }
//     }

//     slice_to_num(&result)
// }

// fn slice_to_num(slice: &BitVec) -> usize {
//     slice.iter().rev().enumerate().fold(0, |r, (i, b)| {
//         r + (2_usize.pow(i as u32)) * (if *b { 1 } else { 0 })
//     })
// }

// fn parse_hex(input: &str) -> usize {
//     usize::from_str_radix(input, 16).ok().unwrap()
// }

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        assert_eq!(part_one("D2FE28"), 6);
        assert_eq!(part_one("8A004A801A8002F478"), 16);
        assert_eq!(part_one("620080001611562C8802118E34"), 12);
        assert_eq!(part_one("C0015000016115A2E0802F182340"), 23);
        assert_eq!(part_one("A0016C880162017C3686B18A3D4780"), 31);
    }
}

File: 2021/day-16/main.py
def parse_input(input) -> str:
    return ''.join([bin(int(ch, 16))[2:].zfill(4) for ch in input])

def part_one(packet: str, versions: list):
    if packet == len(packet) * '0':
        return

    version = int(packet[:3], 2)
    type_id = int(packet[3:6], 2)

    versions.append(version)

    if type_id == 4: # handle literal
        i = 0
        literal_parts = []

        while True:
            literal_part = packet[6 + i * 5:6 + i * 5 + 5]

            literal_parts.append(literal_part)

            if literal_part[0] == "0":
                break

            i += 1

        part_one(packet[6 + i * 5 + 5:], versions)
    elif packet[6] == "0": # handle sub packets of length 15
        length = int(packet[7:22], 2)

        # part_one(packet[22:22+length], versions)
        part_one(packet[22:], versions)
    else: # handle count of sub packets
        sub_count = int(packet[7:18], 2)

        part_one(packet[18:], versions)



# 011 000 1 00000000010
#                      000 000 0 000000000010110 0001000101010110001011001000100000000010000100011000111000110100

# Run part 
input = open('src/input').readline()
versions = []

part_one(parse_input(input), versions)

print("Result", sum(versions))

# Run part one tests
# test_cases = [
#     ("620080001611562C8802118E34", 12),
#     ("D2FE28", 6),
#     ("8A004A801A8002F478", 16),
#     ("C0015000016115A2E0802F182340", 23),
#     ("A0016C880162017C3686B18A3D4780", 31),
# ]

# for (packet, expected_sum) in test_cases:
#     print("Testing", packet)

#     versions = []

#     part_one(parse_input(packet), versions)

#     if sum(versions) != expected_sum:
#         print("Failed on", packet, "Expected", expected_sum, "Got", sum(versions))
#         break